drivers: i2c: add i2c controller implementation using rp2040 pio

yiding opened the pull request:
I2C controller that uses PIO on RP2040. Using this we can drive up to 8 I2C busses in addition to the two supported by the built-in i2c peripheral.

This driver is adapted from the pico sdk pio i2c example with some changes to fix issues identified over the years.

1. Resetting the input shift register before reading a byte. Sometimes the shift register might get out of sync.
2. Resetting the SM after handling a NAK, related to shift registers counters getting out of sync.
3. Waiting for possible clock stretching when writing the I2C STOP.

This is a blocking driver as I found versions interacting with the PIO via interrupts to perform much worse than simply waiting via `k_yield`.

The driver attempts to load the PIO code only once in order to allow multiple driver instances to run the same code.

It also uses the `reg` property as the SM, instead of relying on pico sdk to arbitrarily assign one, which seems to be the norm for other PIO-based drivers in the codebase. I'll put up a separate PR to do this on the main pio dts nodes and fix-up drivers if people think this is a good idea.

yiding commented:
I have a followup which adds support for using a DMA channel to improve performance, which roughly doubles the performance. I'll put that up as a separate PR.

I have also tried implementing this driver using interrupts in two ways, both of which performs worse than this version. 
I made sure to keep the interrupt code in memory using __ramfunc and ld relocations for library code, and it still ended up much slower. 

1. sleep on a semaphore where the code would currently yield, and using an interrupt handler to unblock the thread when condition is met.
2. rtio api implementation that uses state machine framework and all code living in interrupts.

If interested I can dig them up and post them elsewhere.

FlorianWeber1018 commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	rc = 0;
+done:
+	k_mutex_unlock(&data->lock);
+	return 0;
```
```suggestion
	return rc;
```
else a NACK does not lead to an error (tested at a rp2040 with a wrong slave address configured )

teburd commented:
I'm not at all familiar with the pio peripheral, so I've reassigned.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	data->config = I2C_SPEED_SET(I2C_SPEED_DT);
+
+	PIO pio = pio_rpi_pico_get_pio(cfg->piodev);
+	uint sm = cfg->sm;
```
I think this alias is verbose.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	}
+	offset = i2c_pio_program_offset[pionum];
+	k_mutex_unlock(&i2c_pio_program_loader_mtx);
+	return offset;
```
It is common to put a blank line before the return.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	k_mutex_init(&data->lock);
+	data->config = I2C_SPEED_SET(I2C_SPEED_DT);
+
+	PIO pio = pio_rpi_pico_get_pio(cfg->piodev);
```
If you have adopted the old C-style variable declaration at the beginning policy, you should make it consistent throughout.

yiding replied:
No, I don't like that. Most are declared in the block they are used.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	.recover_bus = i2c_pio_recover_bus,
+};
+
+#if NUM_PIOS == 2
```
If this condition varies by instance, it should be defined in a macro that provides a per-instance implementation.

yiding replied:
It does not, this is guarding program-wide static data. The condition is actually tied to the MCU being used. RP2040 has 2 PIO, RP2350 has 3 PIO.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	return 0;
+}
+
+static const struct i2c_driver_api api = {
```
use DEVICE_API

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+static void pio_i2c_wait_idle(PIO pio, uint sm)
+{
+	/* Finished when TX runs dry or SM hits an IRQ */
+	pio->fdebug = 1u << (PIO_FDEBUG_TXSTALL_LSB + sm);
```
Use BIT()

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+		k_yield();
+	}
+	/* some versions of GCC dislike this */
+#ifdef __GNUC__
```
Is this really necessary?

yiding replied:
Some of this is copy pasta'd from the i2c pio example code supplied by rpi. I'm happy to delete this until someone raises a need for it.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+ * Enable autopush to RX fifo. The program is always reading, but if autopush is
+ * disabled the reads are not pushed out.
+ */
+static void pio_i2c_rx_enable(PIO pio, uint sm, bool en)
```
inline

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+		return -ENODEV;
+	}
+
+	__ASSERT(!is_spin_locked(spin_lock_instance(PICO_SPINLOCK_ID_HARDWARE_CLAIM)),
```
Since SMP is not implemented yet, the policy on how to handle hardware spinlock is not decided yet. It is better not to have it for now.

yiding replied:
This was put in here because i noticed when i soft-reset the pico sometimes the spin-lock state isn't reset. Probably can remove this here and ensure spinlock state is reset on startup.

soburi replied:
OK. I got it.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+		}
+		k_yield();
+	}
+	if (pio_i2c_check_error(pio, sm)) {
```
Please leave one blank line after the condition phrase.

soburi reviewed (changes requested):
Sorry for late response. I've added some comments

yiding commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+};
+
+#if NUM_PIOS == 2
+static uint8_t i2c_pio_program_offset[NUM_PIOS] = {255, 255};
```
This is meant to be shared across multiple instances in order to avoid loading the pio program once per instance.

soburi commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	while (len && !pio_i2c_check_error(pio, sm)) {
+		if (!pio_sm_is_tx_fifo_full(pio, sm)) {
+			--len;
+			pio_i2c_put_or_err(pio, sm,
+					   (*txbuf++ << PIO_I2C_DATA_LSB) |
+						   ((len == 0) << PIO_I2C_FINAL_LSB) | 1u);
```
Avoid calling functions and manipulating variables on the same line.

yiding commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,567 @@
+	I2C_SC0_SD0,              /* Pull clock low */
+};
+
+static bool pio_i2c_check_error(PIO pio, uint sm)
```
Function is static, modern compilers makes good inlining decisions without the hint. Same with the other small functions here.

yiding commented:
Updated with additional fixes for handling "gather" style writes (e.g. as used by i2c_burst_write). @FlorianWeber1018

soburi reviewed (dismissed):
LGTM

kartben commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,602 @@
+{
+	struct i2c_pio_data *data = dev->data;
+
+	return data->config;
```
You should return an int and actually set *config

kartben reviewed (commented):
Actually spotted an issue after the fact

kartben reviewed (changes requested):
See above

kartben commented on drivers/i2c/i2c_rpi_pico_pio.c:
```diff
@@ -0,0 +1,602 @@
+};
+
+/**
+ * @name I2C PIO Program
+ * Based on pico example code:
+ * https://github.com/raspberrypi/pico-examples/blob/master/pio/i2c/i2c.pio
```
this program is actually BSD-3 licensed -- you can't just take it and implicitly relicense it as Apache 2. This probably needs some kind of TSC approval, or maybe PIO based driver should be in a dedicated module or something?

yiding replied:
Hm. Originally had the original license in this file but the linters didn't like that at all, so I meant to ask what should be done in this situation, but forgot about it. The C code interfacing with the PIO code is also based on the corresponding C code in the example as well, so if we want to be really diligent about it that should also be licensed BSD3. 

Is it not possible to have BSD3 code in tree? I thought BSD3 is compatible with Apache2, as long as we include the original license somewhere.

soburi replied:
Since the BSD-3 source code already exists, it seems like there should be no problem as long as the license is stated correctly.

https://github.com/zephyrproject-rtos/zephyr/blob/c544e4d08b43d0bcbfa22e380a899d804279fd6a/drivers/sensor/bosch/bmi08x/bmi08x_config_file.h#L5

soburi replied:
@yiding 

Could you put just the BSD-3 part into a separate file? (I think it would be fine to just make it a local header file.

The BSD-3 code already exists, so I think it would be best to explain it to TSC and include it as BSD-3.

yiding replied:
I don't think it would appease the license checker any more than this would.

I'll ask in this discord what's the best way forward here is. The docs are not clear.

soburi replied:
Understood. I'll follow their policy.

henrikbrixandersen replied:
"Importing code into the Zephyr OS from other projects that use a license other than the Apache 2.0 license needs to be fully understood in context and approved by the Zephyr governing board.":

https://docs.zephyrproject.org/latest/contribute/guidelines.html#licensing

yiding committed:
drivers: i2c: add i2c controller implementation using rp2040 pio

I2C controller that uses PIO on RP2040. Using this we can drive up to 8 I2C
busses in addition to the two supported by the built-in i2c peripheral.

This driver is adapted from the pico sdk pio i2c example with some changes
to fix identified issues.

1. Resetting the input shift register before reading a byte. Sometimes the
   shift register might get out of sync.
2. Resetting the SM after handling a NAK, related to shift registers
   counters getting out of sync.
3. Waiting for possible clock stretching when writing the I2C STOP.

This is a blocking driver as I found versions interacting with the PIO via
interrupts to perform much worse than simply waiting via `k_yield`.

The driver attempts to load the PIO code only once in order to allow
multiple driver instances to run the same code.

It also uses the `reg` property as the SM, instead of relying on pico sdk
to arbitrarily assign one, which seems to be the norm for other PIO-based
drivers in the codebase. I'll put up a separate PR to do this on the main
pio dts nodes and fix-up drivers if people think this is a good idea.

Signed-off-by: Yiding Jia <yiding.jia@gmail.com>

yiding commented:
Scancode license checker gives this error, so it seems I can't just trivially declare the license of this file to BSD-3:

```
Error: * scan/drivers/i2c/i2c_rpi_pico_pio.c has invalid license: bsd-new
```

yiding commented:
I created an issue #87378 per documentation. This isn't being contributed as an external component, but it's the closest process I've found for this situation.

carlescufi commented:
@yiding is it only the assembly code that you took from the RPi Pico SDK? Or some of the C code as well? If it's only the PIO asm code, then I suggest you move that to https://github.com/zephyrproject-rtos/hal_rpi_pico instead, and then the driver could be just Apache v2 licensed.
